### PR TITLE
Added VerbInterceptor

### PR DESCRIPTION
**Pull request by kurtaj**
Adrian,

I'm new to github, so let me know if I'm not going about this the right way.  We've been using CLAP for a while and like it a lot.  We ran into issues where we needed instrumentation to tell us what verbs were invoked.

To support this, we added VerbInterceptorAttribute and supporting code to allow the application to log activity.

Please consider these changes and let me know if you have any questions or concerns.

Thanks,
Kurt Jacobson
CSG International

### CODE REVIEW

File: CLAP/Parser.cs [hunk 1]
```diff
@@ -775,6 +791,12 @@ namespace CLAP
             return verbMethods;
         }
 
+        private Method GetMethod(string verb)
```
Comment (adrianaisemberg) [inline CLAP/Parser.cs]: This new method is never used.

File: CLAP/IVerbInvocationInterceptor.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+﻿using System.Collections.Generic;
+
+namespace CLAP
+{
+    public interface IVerbInvocation
```
Comment (adrianaisemberg) [inline CLAP/IVerbInvocationInterceptor.cs]: Please document the interface (xmldoc).

File: CLAP/IVerbInvocationInterceptor.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        public string Verb { get; set; }
+        public Method VerbMethod { get; set; }
+        public List<object> VerbParameters { get; set; }
+        public object TargetObject { get; set; }
+        public Dictionary<string, string> InputArgs { get; set; }
+
+        public void Proceed()
+        {
+            VerbMethod.MethodInfo.Invoke(TargetObject, VerbParameters.ToArray());
```
Comment (adrianaisemberg) [inline CLAP/IVerbInvocationInterceptor.cs]: Please assert that VerbMethod.MethodInfo and VerbParameters are not null.
See my comment regarding VerbParameters above. This property should never be null.

File: CLAP/VerbInterceptorAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+namespace CLAP
+{
+    [Serializable]
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+    public sealed class VerbInterceptorAttribute : Attribute
```
Comment (adrianaisemberg) [inline CLAP/VerbInterceptorAttribute.cs:10]: Please xmldoc the attribute.

File: CLAP/Method.cs [hunk 5]
```diff
@@ -53,6 +53,8 @@ namespace CLAP
             MethodInfo = method;
 
             var verbAttribute = method.GetAttribute<VerbAttribute>();
+            if (verbAttribute == null)
+                return;
```
Comment (adrianaisemberg) [inline CLAP/Method.cs]: Please add { } and new lines. This is the standard all over the solution:
if ()
{
  // even for single-line statements
}

File: CLAP/Method.cs [hunk 6]
```diff
@@ -7,7 +7,7 @@ namespace CLAP
     /// <summary>
     /// A method descriptor
     /// </summary>
-    internal sealed class Method
+    public sealed class Method
```
Comment (adrianaisemberg) [inline CLAP/Method.cs:10]: What is the motivation in this change?
I tried changing it back to internal and everything compiles.
If there is no real motivation - please change it back to internal. Thanks :)

File: CLAP/Parser.cs [hunk 7]
```diff
@@ -81,180 +81,196 @@ namespace CLAP
...
+                    validator.Validate(parametersAndValues.ToArray());
                 }
             }
+
+            if (inputArgs.Any())
+            {
+                throw new UnhandledParametersException(inputArgs);
+            }
+
+            var invocation = new DefaultVerbInvocation() { Verb = verb, TargetObject = obj, VerbMethod = method, VerbParameters = parameters, InputArgs = inputArgs };
```
Comment (adrianaisemberg) [inline CLAP/Parser.cs]: Please break the lines. The standard is to have a single prop each line.
The motivation for this is so future diff's are easy to read.
If one adds, changes or removes a property - only that prop will be highlighted instead of all the line which contains all the props.

new DefaultVerbInvocation
{
    Verb = verb,
    ...
}

Please notice that "()" is not required for the ctor when using this syntax.

File: CLAP/Parser.cs [hunk 8]
```diff
@@ -81,180 +81,196 @@ namespace CLAP
...
-                // find the method by the given verb
-                //
-                var typeVerbs = GetVerbs();
-
-                var method = typeVerbs.FirstOrDefault(v => v.Names.Contains(verb.ToLowerInvariant()));
-
-                // if no method is found - a default must exist
-                //
-                if (method == null)
+                var invocation = BuildVerbInvocation(verb, obj, args);
```
Comment (adrianaisemberg) [inline CLAP/Parser.cs]: Please assert that invocation is not null after this line.

File: CLAP/IVerbInvocationInterceptor.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        string Verb { get; set; }
+        Method VerbMethod { get; set; }
+        List<object> VerbParameters { get; set; }
+        object TargetObject { get; set; }
+        Dictionary<string, string> InputArgs { get; set; }
+        void Proceed();
+    }
+
+
+    public class DefaultVerbInvocation : IVerbInvocation
```
Comment (adrianaisemberg) [inline CLAP/IVerbInvocationInterceptor.cs]: This class can be internal, and if is - also sealed.

File: CLAP/Parser.cs [hunk 10]
```diff
@@ -81,180 +81,196 @@ namespace CLAP
...
+            }
+
+            var invocation = new DefaultVerbInvocation() { Verb = verb, TargetObject = obj, VerbMethod = method, VerbParameters = parameters, InputArgs = inputArgs };
+            return invocation;
+        }
+
+        private Method GetVerbInterceptor()
+        {
+            var method = typeof(T).GetMethodsWith<VerbInterceptorAttribute>().FirstOrDefault();
+            return method == null ? null : new Method(method);
```
Comment (adrianaisemberg) [inline CLAP/Parser.cs]: The only use of the returned value of this method is the MethodInfo property of it.
I suggest returning a MethodInfo (instead of Method) and invoking it directly.
Having this fix, we can revert the changes in the Method class (the "if (verbAttribute == null)" condition).

File: CLAP/Parser.cs [hunk 11]
```diff
@@ -81,180 +81,196 @@ namespace CLAP
...
-                    {
-                        throw new MissingVerbException(verb);
-                    }
-
-                    method = typeVerbs.FirstOrDefault(v => v.IsDefault);
+                    invocation.Proceed();
+                }
+                else
+                {
+                    interceptor.MethodInfo.Invoke(obj, new object[] { invocation });
```
Comment (adrianaisemberg) [inline CLAP/Parser.cs]: Please see my comment in the GetVerbInterceptor method.
After changing it to return a MethodInfo, we can run:
"interceptor.Invoke(" instead of "interceptor.MethodInfo.Invoke("

File: CLAP/Parser.cs (hunk 10 above)
Comment (adrianaisemberg) [inline CLAP/Parser.cs]: I understand that only one VerbInterceptor is supported and any additional will be ignored.
When using reflection to load methods - we can never know the order they will be returned. Taking only the first and ignoring the rest may lead to unpredictable behavior of the app.
I suggest checking for at most one interceptor. If more exist - throw an exception.
Please do this validation in the Validate method. This method is called in the Parser's ctor (so it will fail-fast).
In-addition, please add a special exception type for it.
Please don't forget to add unit tests to cover all these scenarios.

File: CLAP/IVerbInvocationInterceptor.cs [hunk 12]
```diff
@@ -0,0 +1,29 @@
+﻿using System.Collections.Generic;
+
+namespace CLAP
+{
+    public interface IVerbInvocation
+    {
+        string Verb { get; set; }
```
Comment (adrianaisemberg) [inline CLAP/IVerbInvocationInterceptor.cs]: Please remove all setters and keep only getters.
Please correct me if I'm wrong: shouldn't instances of this interface be immutable?
I mean, when an invocation occurs, do we want to allow the invoked method to modify these values?

In-addition, please change the List to a readonly list. We don't want the invoked method to add or remove args.
The same for the dictionary.

If you change them all the getters, please create an instance of the list and dict in the ctor of the default implementation so they are never null.

File: CLAP/Parser.cs [hunk 13]
```diff
@@ -81,180 +81,196 @@ namespace CLAP
...
-                    // if there is a verb input but no method was found - throw
-                    if (verb != null && !s_prefixes.Any(p => p.Equals(verb[0].ToString())))
-                    {
-                        throw new MissingVerbException(verb);
-                    }
-
-                    method = typeVerbs.FirstOrDefault(v => v.IsDefault);
+                    invocation.Proceed();
+                }
+                else
```
Comment (adrianaisemberg) [inline CLAP/Parser.cs]: Unfortunately, there is a bug here.
If there is a method interceptor - the actual verb will never be executed :)
Please see my comment in the unit test you added regarding this.

File: Tests/Tests.cs [hunk 14]
```diff
@@ -1368,5 +1368,20 @@ namespace Tests
...
+        {
+            const string method1 = "foo1";
+            const string method2 = "foo2";
+
+            var sample = new InterceptorSample();
+
+            Parser.Run(new[] { method1 }, sample);
+            Parser.Run(new[] { method2 }, sample);
+
+            Assert.AreEqual(method1, sample.Invoked[0]);
```
Comment (adrianaisemberg) [inline Tests/Tests.cs]: Please also assert  that all the verbs were actually executed. If you do it - you'll spot the bug.

File: Tests/Samples.cs [hunk 15]
```diff
@@ -716,6 +716,30 @@ namespace Tests
...
+        public void Foo1(string str)
+        { }
+
+        [Verb]
+        public void Foo2(string str)
+        { }
+
+        public IList<string> Invoked { get; private set; }
+
+        [VerbInterceptor]
```
Comment (adrianaisemberg) [inline Tests/Samples.cs:836]: Please add a test that makes sure that interception works also in inheritance.
This requires at least two additional tests:
1. A base class with a verb and a derived class with an interceptor.
2. A base with an interceptor and a derived with a verb.
Both should be supported. Please correct me if you think otherwise.

**Comment by adrianaisemberg** [comment]
Kurt, thanks for the idea and for implementing it. I love it!

In addition to defining interceptors using attributes (which is awesome!) - please add the option to define such also with a method such as with RegisterHelpHandler etc.
For example:
var p = new Parser&lt;T>();
p.RegisterInterceptor(Action&lt;IVerbInvocation> interceptor);

The motivation with having all these extra "RegisterXXX" methods is to allow one to use all these features when the verbs are defined in an external class where all we have is the dll and the class is sealed.
Having a sealed class like that - we can't add any interceptors (or help, error and global handlers).
With all these RegisterXXX methods - we can.

If and when you implement this feature - please make sure the method is not called more than once. If we don't allow more than one [VerbInterceptor] than we can't allow more than one registered as such.

Thanks,
Adrian

**Comment by adrianaisemberg** [comment]
Coolness!

I added some comments in github. I hope it sends you an email about it.

Thanks,
Adrian

Adrian Aisemberg
vercue.com http://www.vercue.com
sharpregion.com http://www.sharpregion.com

File: CLAP/Parser.cs (hunk 13 above)
Comment (adrianaisemberg) [inline CLAP/Parser.cs]: Hmm,
But maybe this isn't a bug.
Is the intention to allow an interceptor to cancel an execution of a verb?
I mean, having an interceptor means that only the interceptor is responsible for calling Proceed() ?
If this is the case then sorry, there isn't a bug.

Comment (adrianaisemberg) [inline CLAP/Parser.cs]: May I suggest a little different design:

[PreInterception]
void Foo(IVerbInvocation invocation)
{
   // so some checks
   if (something)
   {
      invocation.Cancel();
   }
}

[PostInterception]
void Bar(IVerbInvocation invocation)
{
}
1. Add a "void Cancel()" method to IVerbInvocation. Add also a "bool Executed" property.
2. Before invoking the verb, if there is a pre-invocation defined - call it.
3. If Cancel wasn't raised - execute the verb and set Executed to true.
4. If there is a post-invocation defined - call it

**Comment by adrianaisemberg** [comment]
In case you consider my suggestion (having a pre and post interceptors), then please consider this change in the Register method:

p.RegisterPreInterceptor(Action&lt;IVerbInvocation> interception);
p.RegisterPostInterceptor(Action&lt;IVerbInvocation> interception);

File: CLAP/Method.cs (hunk 6 above)
Comment (kurtaj) [inline CLAP/Method.cs:10]: Method is a member of the public interface IVerbInvocation, so it needs to be public.

-----Original Message-----

File: CLAP/Parser.cs (hunk 13 above)
Comment (kurtaj) [inline CLAP/Parser.cs]: Well, there was a bug... it was just in the test class.

-----Original Message-----

**Comment by kurtaj** [comment]
Adrian,

Thanks for all the comments!  I've committed an update based on most of them.

I haven't added Pre/Post interceptors at this time because I believe the existing approach achieves the same thing with less complexity: the interceptor can perform activities before and after calling Proceed().  Obviously, it's your project so do whatever you feel is best.  We'll adapt :-)

As I said, I'm new to github so I'm not sure what I should do next... should I close the existing pull request and send a new one, or will you be able to take from my fork with the existing request?

Thanks,
Kurt

-----Original Message-----

**Comment by adrianaisemberg** [comment]
Hi Kurt,

I implemented the interception feature (thanks again!).
I will publish the new feature in the project's web site and send it to you
once it is ready.
I'll be happy to hear some impressions, ideas and your thoughts on my
implementation.

Meanwhile you can update from the master branch and get the new code.

Thanks,
Adrian

Adrian Aisemberg
vercue.com http://www.vercue.com
sharpregion.com http://www.sharpregion.com

**Comment by kurtaj** [comment]
Thanks!  I noticed the change, but haven't taken a good look at it yet.

Let me tell you a little about how we're using CLAP, so you can get an idea why I made some of the more recent changes on my fork.

I inherited a prototype project that uses CLAP in two phases:
1. First phase uses CLAP to parse the command line and execute a verb, just like you'd expect.
  a. There are currently 2 verbs that may be invoked from the command line.  Each verb loads a different set of modules and behaviors.  These modules contain an additional set of verbs.
  b. Another CLAP parser is created for the loaded module.
2. The second CLAP parser is then invoked using any arguments from the command line that weren't used by the original verb.

I'll take a look at your interceptor implementation in the next few days and let you know what I think.  Thanks for accepting the changes!

-Kurt

-----Original Message-----

**Comment by adrianaisemberg** [comment]
Hi Kurt,

I am a little confused by the second parser.
You said it uses any arguments that were *not *used by the first parser.
However, if there is an argument that is not handled by a parser - an
exception is thrown.

Can you please elaborate a little more on that?

Thanks,
Adrian

Adrian Aisemberg
vercue.com http://www.vercue.com
sharpregion.com http://www.sharpregion.com

**Comment by kurtaj** [comment]
Right, I had to change things around so that the exception is only thrown if there is no interceptor.

I'm guessing when the prototype was written a couple months ago, CLAP didn't have that piece of validation.

-Kurt

-----Original Message-----

**Comment by adrianaisemberg** [comment]
I see.
It makes sense now.

Thanks for the update.

Adrian Aisemberg
vercue.com http://www.vercue.com
sharpregion.com http://www.sharpregion.com
